### PR TITLE
Custom parsers

### PR DESCRIPTION
**Pull request by d-biehl**
This is the result of our RoboCon sprint to define and implement the custom parser API see #1283

What's missing at moment are the unit test cases for that API.

We need to talk about the Documentation statement value property. Can call the the from_params with a mutliline text with 2 paragraphs, but the value property deletes `\n' characters, so I get only one paragraph. Is this correct?

### CODE REVIEW

**Comment by pekkaklarck** [comment]
Thanks, I hopefully have time to look at this during the weekend or latest next week. Some quick comments already now:

- I didn't fully understand the issue with `Documentation.from_params` but there certainly can be bugs. That probably deserves its own issue.
- As we discussed at the sprints, the parser API itself will be relatively simple with just few methods like `def parse_suite_file(path: Path, defaults: Defaults)`. That then makes our `Default` object part of the public API and it definitely needs some cleanup. I can take care of that myself.
- Same tests related to reStructuredText syntax are failing on CI. Most likely the parser changes have affected the reST parser. Can you take a look?

**Comment by d-biehl** [comment]
created the issue #4670 to fix the Documentation thing independently.

next step is finding out how custom resource files can be parsed. Seems a little bit complication because they where imported by a static IMPORTER class in the namespace. @pekkaklarck do you have a quick idea?

File: src/robot/running/builder/parsers.py [hunk 1]
```diff
@@ -34,12 +38,25 @@ def parse_suite_file(self, source: Path, defaults: Defaults = None):
...
+    @property
+    def extensions(self):
+        raise NotImplementedError
+
+    @property
+    def resource_extensions(self):
+        raise NotImplementedError
+
+    def _get_curdir(self, source):
+        return str(source.parent).replace('\\', '\\\\') if self.process_curdir else None
```
Comment (pekkaklarck) [inline src/robot/running/builder/parsers.py:50]: If we want this to be part of the public API, it shouldn't have `_`.

**Comment by pekkaklarck** [comment]
I commented this PR already some time ago and submitted some design decision question to the [original issue](https://github.com/robotframework/robotframework/issues/1283#issuecomment-1508212701) a week ago. There were no comments from anyone, so I decided to go with the design I had proposed. The initial implementation is now ready (d973fb86b34f7290ccc2377889415698d527f3c6) and it makes this PR obsolete. There's still some [work to do before this feature is ready](https://github.com/robotframework/robotframework/issues/1283#issuecomment-1517705142), but I plan to get all that done early next week. If you @d-biehl have a change to test the new API it would be great!

**Pull request closed without merging**